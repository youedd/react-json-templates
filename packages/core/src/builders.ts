import { type FIXME, type RJT_CONSTANT_TYPE, type RJT_ACTION, type RJT_CONSTANT, type RJT_OPERATION, type RJT_OPERATION_TYPE, type RJT_COMPONENT, type RJT_PROPS, type RJT_ELEMENT, type RJT_FRAGMENT } from './types'

export const Serializable = <T>(_name: string, Comp: T): T => Comp

export const Component = (name: string, props: RJT_PROPS): RJT_COMPONENT => {
  return {
    type: '__RJT_COMPONENT__',
    name,
    props
  }
}

export const Fragment = (children: RJT_ELEMENT[]): RJT_FRAGMENT => {
  return {
    type: '__RJT_FRAGMENT__',
    children
  }
}

export const Action = (name: string, ...params: unknown[]): RJT_ACTION => {
  return {
    type: '__RJT_ACTION__',
    name,
    params
  }
}

export const Constant = <T extends RJT_CONSTANT_TYPE>(name: string, type: T): RJT_CONSTANT<T> => {
  return {
    type: '__RJT_CONSTANT__',
    name,
    kind: type
  }
}

export const Operation = <T extends RJT_OPERATION_TYPE>(
  operation: T,
  ...operands: RJT_OPERATION<T>['operands']
): FIXME => {
  return {
    type: '__RJT_OPERATION__',
    operation,
    operands
  }
}
