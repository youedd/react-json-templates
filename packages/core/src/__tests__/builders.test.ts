import * as Utils from '../builders'

describe('utils', () => {
  it('should build a component object', () => {
    expect(Utils.Component('S1', {})).toMatchSnapshot()
    expect(Utils.Component('S1', { for: 'bar' })).toMatchSnapshot()
  })

  it('should build a fragment object', () => {
    expect(Utils.Fragment([])).toMatchSnapshot()
    expect(Utils.Fragment([Utils.Component('S1', {})])).toMatchSnapshot()
  })

  it('should build an action object', () => {
    expect(Utils.Action('GO_TO', 'screen1', { foo: 'bar' })).toMatchSnapshot()
  })

  describe(Utils.Constant, () => {
    it('should build a constant number', () => {
      expect(Utils.Constant('WINDOW_WIDTH', 'number')).toMatchSnapshot()
    })

    it('should build a constant string', () => {
      expect(Utils.Constant('TRANSLATION', 'string')).toMatchSnapshot()
    })

    it('should build a constant function', () => {
      expect(Utils.Constant('HELPER', 'function')).toMatchSnapshot()
    })

    it('should build a constant array', () => {
      expect(Utils.Constant('ITEMS', 'array')).toMatchSnapshot()
    })

    it('should build a constant object', () => {
      expect(Utils.Constant('CONFIG', 'object')).toMatchSnapshot()
    })
  })

  describe(Utils.Operation, () => {
    it('should build an addition operation', () => {
      expect(Utils.Operation('+', 2, 5)).toMatchSnapshot()
    })

    it('should build a subtraction operation', () => {
      expect(Utils.Operation('-', 2, 5)).toMatchSnapshot()
    })

    it('should build a division operation', () => {
      expect(Utils.Operation('/', 2, 5)).toMatchSnapshot()
    })

    it('should build a multiplication operation', () => {
      expect(Utils.Operation('*', 2, 5)).toMatchSnapshot()
    })

    it('should build a pow operation', () => {
      expect(Utils.Operation('pow', 2, 5)).toMatchSnapshot()
    })

    it('should build a sqrt operation', () => {
      expect(Utils.Operation('sqrt', 2)).toMatchSnapshot()
    })

    it('should build an operation with constant operand', () => {
      expect(Utils.Operation('*', 2, Utils.Constant('MIN_WIDTH', 'number'))).toMatchSnapshot()
    })

    it('should build nested operations', () => {
      expect(Utils.Operation('*', Utils.Operation('sqrt', 2), Utils.Constant('MIN_WIDTH', 'number'))).toMatchSnapshot()
    })
  })
})
