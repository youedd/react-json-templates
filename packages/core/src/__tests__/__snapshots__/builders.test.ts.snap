// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils Constant should build a constant array 1`] = `
{
  "kind": "array",
  "name": "ITEMS",
  "type": "__RJT_CONSTANT__",
}
`;

exports[`utils Constant should build a constant function 1`] = `
{
  "kind": "function",
  "name": "HELPER",
  "type": "__RJT_CONSTANT__",
}
`;

exports[`utils Constant should build a constant number 1`] = `
{
  "kind": "number",
  "name": "WINDOW_WIDTH",
  "type": "__RJT_CONSTANT__",
}
`;

exports[`utils Constant should build a constant object 1`] = `
{
  "kind": "object",
  "name": "CONFIG",
  "type": "__RJT_CONSTANT__",
}
`;

exports[`utils Constant should build a constant string 1`] = `
{
  "kind": "string",
  "name": "TRANSLATION",
  "type": "__RJT_CONSTANT__",
}
`;

exports[`utils Operation should build a division operation 1`] = `
{
  "operands": [
    2,
    5,
  ],
  "operation": "/",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build a multiplication operation 1`] = `
{
  "operands": [
    2,
    5,
  ],
  "operation": "*",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build a pow operation 1`] = `
{
  "operands": [
    2,
    5,
  ],
  "operation": "pow",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build a sqrt operation 1`] = `
{
  "operands": [
    2,
  ],
  "operation": "sqrt",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build a subtraction operation 1`] = `
{
  "operands": [
    2,
    5,
  ],
  "operation": "-",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build an addition operation 1`] = `
{
  "operands": [
    2,
    5,
  ],
  "operation": "+",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build an operation with constant operand 1`] = `
{
  "operands": [
    2,
    {
      "kind": "number",
      "name": "MIN_WIDTH",
      "type": "__RJT_CONSTANT__",
    },
  ],
  "operation": "*",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils Operation should build nested operations 1`] = `
{
  "operands": [
    {
      "operands": [
        2,
      ],
      "operation": "sqrt",
      "type": "__RJT_OPERATION__",
    },
    {
      "kind": "number",
      "name": "MIN_WIDTH",
      "type": "__RJT_CONSTANT__",
    },
  ],
  "operation": "*",
  "type": "__RJT_OPERATION__",
}
`;

exports[`utils should build a component object 1`] = `
{
  "name": "S1",
  "props": {},
  "type": "__RJT_COMPONENT__",
}
`;

exports[`utils should build a component object 2`] = `
{
  "name": "S1",
  "props": {
    "for": "bar",
  },
  "type": "__RJT_COMPONENT__",
}
`;

exports[`utils should build a fragment object 1`] = `
{
  "children": [],
  "type": "__RJT_FRAGMENT__",
}
`;

exports[`utils should build a fragment object 2`] = `
{
  "children": [
    {
      "name": "S1",
      "props": {},
      "type": "__RJT_COMPONENT__",
    },
  ],
  "type": "__RJT_FRAGMENT__",
}
`;

exports[`utils should build an action object 1`] = `
{
  "name": "GO_TO",
  "params": [
    "screen1",
    {
      "foo": "bar",
    },
  ],
  "type": "__RJT_ACTION__",
}
`;
